### PR TITLE
Added check for integer overflow in get_num_images 

### PR DESCRIPTION
**Pull request by Eharve14**
As discussed in pull request 1396, added a check for integer overflow. 
Change list:
Defined num_images as unsigned int
Moved the if statement to check for an empty directory to the beginning of the read directory section
Added a check to see if num images rolls back to zero when incrementing.

### CODE REVIEW

File: src/bin/jp2/opj_compress.c [hunk 1]
```diff
@@ -486,6 +486,9 @@ static unsigned int get_num_images(char *imgdirpath)
             continue;
         }
         num_images++;
+        if (num_images == 0) {
```
Comment (rouault) [inline src/bin/jp2/opj_compress.c:489]: while this is technically correct, having behaviour that relies on overflow of unsigned types always makes me a bit nervous. I'd prefer a test on UINT_MAX before the incrementation is done (the advantage is that a build with  clang -fsanitize=undefined,unsigned-integer-overflow would also be happy). Same for below similar instances

File: src/bin/jp2/opj_decompress.c [hunk 2]
```diff
@@ -374,7 +374,7 @@ int get_num_images(char *imgdirpath)
 {
     DIR *dir;
     struct dirent* content;
-    int num_images = 0;
+    unsigned int num_images = 0;
```
[the remark this answers is not in the document]
Comment (Eharve14) [reply src/bin/jp2/opj_decompress.c:377]: Okay, I will updated it.

File: src/bin/jp2/opj_compress.c (hunk 1 above)
Comment (Eharve14) [reply src/bin/jp2/opj_compress.c]: Revised as requested.

**Comment by Eharve14** [comment]
@rouault I made the requested revisions. I was under the impression that unsigned overflow was always defined behavior. Thank you for the feedback.

File: src/bin/jp2/opj_compress.c [hunk 3]
```diff
@@ -485,6 +486,9 @@ static unsigned int get_num_images(char *imgdirpath)
         if (strcmp(".", content->d_name) == 0 || strcmp("..", content->d_name) == 0) {
             continue;
         }
+        if (num_images == UINT_MAX) {
+            fprintf(stderr, "Too many files in folder %s\n", imgdirpath);
```
Comment (rouault) [inline src/bin/jp2/opj_compress.c:490]: missing break;

File: src/bin/jp2/opj_decompress.c [hunk 4]
```diff
@@ -388,7 +389,11 @@ int get_num_images(char *imgdirpath)
         if (strcmp(".", content->d_name) == 0 || strcmp("..", content->d_name) == 0) {
             continue;
         }
+        if (num_images == UINT_MAX) {
+            fprintf(stderr, "Too many files in folder %s\n", imgdirpath);
```
Comment (rouault) [inline src/bin/jp2/opj_decompress.c:393]: missing break;

File: src/bin/jp2/opj_dump.c [hunk 5]
```diff
@@ -140,6 +141,9 @@ static int get_num_images(char *imgdirpath)
         if (strcmp(".", content->d_name) == 0 || strcmp("..", content->d_name) == 0) {
             continue;
         }
+        if (num_images == UINT_MAX) {
+            fprintf(stderr, "Too many files in folder %s\n", imgdirpath);
```
Comment (rouault) [inline src/bin/jp2/opj_dump.c:145]: missing break;

File: src/bin/jp2/opj_compress.c (hunk 3 above)
Comment (Eharve14) [reply src/bin/jp2/opj_compress.c:490]: Added return statement

**Comment by Eharve14** [comment]
Added the return statement after the overflow check. I don't know how I managed to mess up copy paste.

File: src/bin/jp2/opj_compress.c (hunk 3 above)
Comment (rouault) [reply src/bin/jp2/opj_compress.c:490]: if you use return, you need an explicit closedir(dir) before.
Other possibilities do:
num_images = 0;
break;
Same for other instances

Comment (Eharve14) [reply src/bin/jp2/opj_compress.c:490]: Ah got you, for whatever reason I was thinking the goto fin statement in the calling section of code cleaned everything up, including the dir.  Also didn't realize the opendir function closed the directory if there was nothing there and freed the memory. I made the correction as requested. Thanks again.

**Comment by kirotawa** [comment]
hey there, 

Is these fixes anyhow related with CVE-2021-29338 ? Should it have another CVE assigned to these fixes/commits in this merge?

Thanks!

**Pull request merged**